### PR TITLE
Add four_wheel_steering_controller

### PR DESCRIPTION
**Pull request by vincentrou**
This PR depends on this new package : https://github.com/ros-controls/urdf_geometry_parser/pull/2

ref #55

### CODE REVIEW

File: four_wheel_steering_controller/scripts/teleop_four_wheel_steering_joy.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+import rospy
+from math import pi
+from four_wheel_steering_msgs.msg import FourWheelSteering
+from sensor_msgs.msg import Joy
+
+class TeleopFourWheelSteeringJoy():
```
Comment (bmagyar) [inline four_wheel_steering_controller/scripts/teleop_four_wheel_steering_joy.py]: could you please remove this from this PR? let's try to minimise the scope

File: four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2013, PAL Robotics, S.L.
```
Comment (bmagyar) [inline four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h:4]: this is a lot of copied code from the `diff_drive_controller`, could you use that one instead?

File: four_wheel_steering_controller/src/odometry.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+  void Odometry::init(const ros::Time& time)
+  {
+    // Reset accumulators and timestamp:
+    resetAccumulators();
+    timestamp_ = time;
+  }
+
+
+  bool Odometry::update(const double &fl_speed, const double &fr_speed,
```
Comment (bmagyar) [inline four_wheel_steering_controller/src/odometry.cpp]: param setup and update functions are different, but integration and interface is the same as `diff_drive_controller`. Maybe we can make a super class?

File: four_wheel_steering_controller/test/config/view_odom.rviz [hunk 4]
```diff
@@ -0,0 +1,158 @@
+Panels:
```
Comment (bmagyar) [inline four_wheel_steering_controller/test/config/view_odom.rviz]: please don't put this in

**Review by bmagyar — CHANGES_REQUESTED** [review]

File: four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h (hunk 2 above)
Comment (vincentrou) [reply four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h:4]: I agree with you but is it fine to depend on the diff_drive_controller ? Or is it better to create another package with speed_limiter and odometry library ?

File: four_wheel_steering_controller/src/odometry.cpp (hunk 3 above)
Comment (vincentrou) [reply four_wheel_steering_controller/src/odometry.cpp]: I agree but maybe in a new package ? (as said above)

**Comment by MoriKen254** [comment]
Hello. I also worked on steering drive controller stuff and opened pull request (#289) but they suggested I should join here to avoid duplication. Please check information about my current project on #289. We provided sample demonstration application and documentation as well.

I also want to contribute to this pull request for simplification, extension, documentation and so on, if you don't mind.

Now I'm checking `four_wheel_steering_controller` codes to compare with my `steer_drive_controller` and trying to run it in my PC. I believe both projects share similar concept but I found an important difference. 

---

 For instance, the super class of mine is different from yours: my `steer_drive_controller` inherits `MultiInterfaceController` (available from Jade) which initially handles multiple interfaces such as both position and velocity ones, while `four_wheel_steering_controller` does `ControllerBase` and looks resolve multiple interfaces initialization in itself.

So I believe using `MultiInterfaceController` way will avoid code duplication and make the codes simpler.

Like this, I might be able to contribute to something on this request from my struggles on  `steer_drive_controller`.

---

By the way, I also agree with an idea to create a super class (such as `wheels_drive_controller` ?) to avoid copying codes from `diff_drive_controller`. Maybe we should solve this one first. Please let me work on some modifications you haven't started yet.

Thank you.

**Comment by vincentrou** [comment]
I agree the use of `MultiInterfaceController` will simplify the code. But I developed it for indigo and kinetic.
I think this package will be released just for kinetic so we can use the `MultiInterfaceController`.

**Comment by MoriKen254** [comment]
@vincentrou 

Thank you for your reply! That makes sense if it's compatible with indigo. As you mentioned, if you just release it for kinetic `MultiInterfaceController` would be good.

By the way, I've read your codes and found a difference from mine. I posted a detail on #289 so please have a look at it if you are interested.

**Comment by vincentrou** [comment]
As I mentioned in the PR #289 , I don't think it is relevant to merge it with this one.
@bmagyar is it possible to go forward with this PR ?

File: four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h (hunk 2 above)
Comment (bmagyar) [reply four_wheel_steering_controller/include/four_wheel_steering_controller/speed_limiter.h:4]: Maybe better not...I'm a bit hesitant on this: methodically that's the right choice but from a release and maintenance point of view it's better if things are not that entangled. 

@ipa-mdl what do you think?

**Comment by bmagyar** [comment]
I've restarted the tests. Bear in mind that reviewing this amount of code will take a bit of time.

File: four_wheel_steering_controller/src/four_wheel_steering_controller.cpp [hunk 5]
```diff
@@ -220,26 +220,26 @@ namespace four_wheel_steering_controller{
...
         return false;
       else
-        controller_nh.setParam("wheel_radius",wheel_radius_);
+        root_nh.setParam("wheel_radius",wheel_radius_);
     if(lookup_wheel_base)
       if(!uvk.getDistanceBetweenJoints(front_wheel_names[0], rear_wheel_names[0], wheel_base_))
         return false;
       else
-        controller_nh.setParam("wheel_base",wheel_base_);
+        root_nh.setParam("wheel_base",wheel_base_);
```
Comment (bmagyar) [inline four_wheel_steering_controller/src/four_wheel_steering_controller.cpp]: I would remove this commit...`root_nh` here points to global param space, right? 
Those parameters should _especially_ be in a namespaced environment

File: four_wheel_steering_controller/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,65 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(four_wheel_steering_controller)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+
+#Add custom (non compiling) targets so launch scripts and python files show up in QT Creator's project view.
+file(GLOB_RECURSE EXTRA_FILES */*)
```
Comment (bmagyar) [inline four_wheel_steering_controller/CMakeLists.txt]: Please remove these changes, we shouldn't include qtcreator hacks in the cmake of released packages.

File: four_wheel_steering_controller/src/four_wheel_steering_controller.cpp (hunk 5 above)
Comment (vincentrou) [reply four_wheel_steering_controller/src/four_wheel_steering_controller.cpp]: The thing is that I want to use those parameters in other nodes (so they don't have to parse again the urdf).
When the parameters are in the global space, other nodes do not need to know the controller node name to find them...
Maybe I could use service instead.

File: four_wheel_steering_controller/CMakeLists.txt (hunk 6 above)
Comment (vincentrou) [reply four_wheel_steering_controller/CMakeLists.txt]: agreed

File: four_wheel_steering_controller/src/four_wheel_steering_controller.cpp (hunk 5 above)
Comment (bmagyar) [reply four_wheel_steering_controller/src/four_wheel_steering_controller.cpp]: The param server is exactly for this kind of stuff. And there's nothing wrong with having to know the name of the controller node, that's not something you'd change anyways. I usually use `mobile_base_controller` as name for that on most robots regardless of what exact type of base controller that is.

Comment (bmagyar) [reply four_wheel_steering_controller/src/four_wheel_steering_controller.cpp]: Note with researcher hat on: Keep in mind that services for your experiments are very bad as they are not transparent to anything. At least rosparam can be dumped to a file for processing later.

**Comment by bmagyar** [comment]
There's a failing tests that remains: 
```
[four_wheel_steering_controller.rosunit-four_wheel_steering_controller_4ws_cmd_test/testCrab][FAILURE]
/root/catkin_ws/src/ros_controllers/four_wheel_steering_controller/test/src/four_wheel_steering_4ws_cmd_test.cpp:84
The difference between sqrt(dx*dx + dy*dy) and cmd_vel.speed*travel_time is 0.01000576054909641, which exceeds POSITION_TOLERANCE, where sqrt(dx*dx + dy*dy) evaluates to 1.5899942632927615, cmd_vel.speed*travel_time evaluates to 1.6000000238418579, and POSITION_TOLERANCE evaluates to 0.01.
```

Just add some eps to account for this and I think we can merge.

**Comment by vincentrou** [comment]
The tests are passing now for four_wheel_steering_controller.
But there is one error with the diff_drive_controller on some builds.
Maybe it is because I do not update my branch...

**Comment by vincentrou** [comment]
Is there any other change requested for this PR ?

**Review by bmagyar — APPROVED** [review]

**Comment by bmagyar** [comment]
It is all ok now, thanks for being so patient and stick all the way through!
Merging...

**Comment by Maik93** [comment]
Hi all,
just cloned `ros_controllers` repo on a fresh Kinetic installation (cloning and switching to `kinetic_devel` branch). When I run `catkin_make` I encounter this error:

> Could not find a package configuration file provided by "four_wheel_steering_msgs" with any of the following names:
> four_wheel_steering_msgsConfig.cmake
> four_wheel_steering_msgs-config.cmake
> ...

Inside `four_wheel_steering` directory, there's no msgs, so I don't know what to do; by now I've removed the entire directory (I don't need this package at the moment), and everything works smoothly, I'm just here to tell you there's something incomplete in this pull request, I believe.

Don't know if open a new issue too, you tell me.

**Comment by bmagyar** [comment]
You simply forgot to install the dependencies. There is nothing incomplete in this pull request.
To make your life easier next time use `rosdep install --from-paths src --ignore-src`.

**Comment by Maik93** [comment]
Happy to hear that everything's ok then. I can confirm that it works now (even though there are some problem due to manifest's encoding if using python 2.7, as pointed out [here](https://github.com/ros-infrastructure/rospkg/issues/154)).

An alternative command to yours would have been this: `rosdep install four_wheel_steering_controller`.

Thanks again for your help and sorry for the misleading issue.

**Pull request merged**